File: python-samples/src/samples/exercise02/aggregator.py
"""
Challenge 2: Refactoring to meet standards.

This module works but is hard to read and maintain:
- One long function that does validation, filtering, transformation and stats
- Unclear names in places (d, c, t, v)
- Mixed concerns; should be split into smaller functions
- Participants refactor with Copilot and run tests after each change.
"""


def process_and_aggregate(data, config):
    """
    Process records and return transformed data plus summary stats.

    Args:
        data: List of dicts with 'value' and optional 'id', 'status'.
        config: Dict with 'min_value', 'multiplier', optional 'status_filter'.

    Returns:
        Dict with 'items' (list of processed items) and 'stats' (count, total, average).
    """
    min_val = config.get("min_value", 0)
    mult = config.get("multiplier", 1.0)
    status_filter = config.get("status_filter")

    if not data or not isinstance(data, list):
        return {"items": [], "stats": {"count": 0, "total": 0.0, "average": 0.0}}

    items = []
    total = 0.0
    count = 0

    for d in data:
        if not isinstance(d, dict) or "value" not in d:
            continue
        v = d["value"]
        if status_filter and d.get("status") != status_filter:
            continue
        if v < min_val:
            continue
        new_v = v * mult
        total += new_v
        count += 1
        items.append({"id": d.get("id", "unknown"), "value": round(new_v, 2)})

    avg = total / count if count > 0 else 0.0
    return {
        "items": items,
        "stats": {"count": count, "total": round(total, 2), "average": round(avg, 2)},
    }


def main():
    """Example usage."""
    data = [
        {"id": "a", "value": 100, "status": "active"},
        {"id": "b", "value": 50, "status": "active"},
        {"id": "c", "value": 200, "status": "pending"},
    ]
    config = {"min_value": 40, "multiplier": 1.1, "status_filter": "active"}
    result = process_and_aggregate(data, config)
    print("Stats:", result["stats"])
    for item in result["items"]:
        print(f"  {item['id']}: {item['value']}")


if __name__ == "__main__":
    main()

File: node-samples/src/exercise02/aggregator.js
/**
 * Challenge 2: Refactoring to meet standards.
 *
 * This module works but is hard to read and maintain:
 * - One long function that does validation, filtering, transformation and stats
 * - Unclear names in places (d, v, new_v)
 * - Mixed concerns; should be split into smaller functions
 * Participants refactor with Copilot and run tests after each change.
 */
import { fileURLToPath } from "node:url";

/**
 * Process records and return transformed data plus summary stats.
 *
 * @param {Array<{id?: string, value: number, status?: string}>} data - Records with value and optional id, status.
 * @param {{min_value?: number, multiplier?: number, status_filter?: string}} config - Config.
 * @returns {{items: Array<{id: string, value: number}>, stats: {count: number, total: number, average: number}}}
 */
export function processAndAggregate(data, config) {
  const minVal = config.min_value ?? 0;
  const mult = config.multiplier ?? 1.0;
  const statusFilter = config.status_filter;

  if (!data || !Array.isArray(data)) {
    return {
      items: [],
      stats: { count: 0, total: 0, average: 0 },
    };
  }

  const items = [];
  let total = 0;
  let count = 0;

  for (const d of data) {
    if (!d || typeof d !== "object" || typeof d.value !== "number") {
      continue;
    }
    const v = d.value;
    if (statusFilter != null && d.status !== statusFilter) {
      continue;
    }
    if (v < minVal) {
      continue;
    }
    const newV = v * mult;
    total += newV;
    count += 1;
    items.push({
      id: d.id ?? "unknown",
      value: Math.round(newV * 100) / 100,
    });
  }

  const avg = count > 0 ? total / count : 0;
  return {
    items,
    stats: {
      count,
      total: Math.round(total * 100) / 100,
      average: Math.round(avg * 100) / 100,
    },
  };
}

export function main() {
  const data = [
    { id: "a", value: 100, status: "active" },
    { id: "b", value: 50, status: "active" },
    { id: "c", value: 200, status: "pending" },
  ];
  const config = {
    min_value: 40,
    multiplier: 1.1,
    status_filter: "active",
  };
  const result = processAndAggregate(data, config);
  console.log("Stats:", result.stats);
  for (const item of result.items) {
    console.log(`  ${item.id}: ${item.value}`);
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) main();